lib/web/apiserver.go

```diff
@@ -283,160 +289,163 @@ func NewHandler(cfg Config, opts ...HandlerOption) (*APIHandler, error) {
-	h.GET("/webapi/mfa/devices", h.WithAuth(h.getMFADevicesHandle))
-	h.POST("/webapi/mfa/authenticatechallenge", h.WithAuth(h.createAuthenticateChallengeHandle))
-	h.POST("/webapi/mfa/devices", h.WithAuth(h.addMFADeviceHandle))
-	h.POST("/webapi/mfa/authenticatechallenge/password", h.WithAuth(h.createAuthenticateChallengeWithPassword))
-
-	// trusted clusters
-	h.POST("/webapi/trustedclusters/validate", httplib.MakeHandler(h.validateTrustedCluster))
-
-	// User Status (used by client to check if user session is valid)
-	h.GET("/webapi/user/status", h.WithAuth(h.getUserStatus))
-
 	// Issue host credentials.
 	h.POST("/webapi/host/credentials", httplib.MakeHandler(h.hostCredentials))
 
-	h.GET("/webapi/roles", h.WithAuth(h.getRolesHandle))
-	h.PUT("/webapi/roles", h.WithAuth(h.upsertRoleHandle))
-	h.POST("/webapi/roles", h.WithAuth(h.upsertRoleHandle))
-	h.DELETE("/webapi/roles/:name", h.WithAuth(h.deleteRole))
+	if !cfg.MinimalReverseTunnelRoutesOnly {
 
-	h.GET("/webapi/github", h.WithAuth(h.getGithubConnectorsHandle))
-	h.PUT("/webapi/github", h.WithAuth(h.upsertGithubConnectorHandle))
-	h.POST("/webapi/github", h.WithAuth(h.upsertGithubConnectorHandle))
-	h.DELETE("/webapi/github/:name", h.WithAuth(h.deleteGithubConnector))
+		// ping endpoint is used to check if the server is up. the /webapi/ping
+		// endpoint returns the default authentication method and configuration that
+		// the server supports. the /webapi/ping/:connector endpoint can be used to
+		// query the authentication configuration for a specific connector.
+		h.GET("/webapi/ping", httplib.MakeHandler(h.ping))
+		h.GET("/webapi/ping/:connector", httplib.MakeHandler(h.pingWithConnector))
 
-	h.GET("/webapi/trustedcluster", h.WithAuth(h.getTrustedClustersHandle))
-	h.PUT("/webapi/trustedcluster", h.WithAuth(h.upsertTrustedClusterHandle))
-	h.POST("/webapi/trustedcluster", h.WithAuth(h.upsertTrustedClusterHandle))
-	h.DELETE("/webapi/trustedcluster/:name", h.WithAuth(h.deleteTrustedCluster))
+		// Unauthenticated access to JWT public keys.
+		h.GET("/.well-known/jwks.json", httplib.MakeHandler(h.jwks))
 
-	h.GET("/webapi/apps/:fqdnHint", h.WithAuth(h.getAppFQDN))
-	h.GET("/webapi/apps/:fqdnHint/:clusterName/:publicAddr", h.WithAuth(h.getAppFQDN))
+		// Unauthenticated access to the message of the day
+		h.GET("/webapi/motd", httplib.MakeHandler(h.motd))
 
-	// Desktop access endpoints.
-	h.GET("/webapi/sites/:site/desktops", h.WithClusterAuth(h.clusterDesktopsGet))
-	h.GET("/webapi/sites/:site/desktops/:desktopName", h.WithClusterAuth(h.getDesktopHandle))
-	// GET /webapi/sites/:site/desktops/:desktopName/connect?access_token=<bearer_token>&username=<username>&width=<width>&height=<height>
-	h.GET("/webapi/sites/:site/desktops/:desktopName/connect", h.WithClusterAuth(h.desktopConnectHandle))
-	// GET /webapi/sites/:site/desktopplayback/:sid?access_token=<bearer_token>
-	h.GET("/webapi/sites/:site/desktopplayback/:sid", h.WithAuth(h.desktopPlaybackHandle))
```

Comment:
Note: I haven't eyeballed every single endpoint, as I think we may refactor this a bit before the end.